File: src/py/polymer.py
import numpy as np

class Polymer(object):

    def __init__(self, N, k_harm, k_F, R0, eps, sigma, omega, gamma, T, dt):
        self.N = N
        self.x = np.arange(0,N,dtype='float')
        self.y = np.ones(N)*N
        self.z = np.zeros(N)
        self.vx = np.zeros(N)
        self.vy = np.zeros(N)
        self.vz = np.zeros(N)
        self.vhx = np.zeros(N)
        self.vhy = np.zeros(N)
        self.vhz = np.zeros(N)
        self.fx = np.zeros(N)
        self.fy = np.zeros(N)
        self.fz = np.zeros(N)
        self.k_harm = k_harm
        self.k_F = k_F
        self.R0 = R0
        self.eps = eps
        self.sigma2 = sigma**2
        self.cutoff2 = 1.1224620483093729**2
        self.omega = omega
        self.cmx = self.cmy = self.cmz = 0.0
        self.rgx = self.rgy = self.rgz = 0.0
               
        self.fx, self.fy, self.fz = self.F_tot()
        self.update_cm()
        self.update_rg()  

        self.dt = dt
        self.T = T 
        self.gamma = gamma
        self.mu = np.sqrt(2.0*gamma*T/self.dt)

        self.rx = np.random.normal(0,1,self.N)*self.mu
        self.ry = np.random.normal(0,1,self.N)*self.mu
        self.rz = np.random.normal(0,1,self.N)*self.mu

    def U_harm(self):
        U = 0.0
        for i in range(0,self.N-1):
            r2 = (self.x[i] - self.x[i+1])**2 + \
                    (self.y[i] - self.y[i+1])**2 + \
                    (self.z[i] - self.z[i+1])**2
            U += 0.5*self.k_harm*r2
        return U

    def U_LJ(self):
        U = 0.0
        for i in range(0,self.N):
            for j in range(i+1,self.N):
                dx = self.x[i] - self.x[j]
                dy = self.y[i] - self.y[j]
                dz = self.z[i] - self.z[j]
                r2 = dx**2 + dy**2 + dz**2
                if r2 < self.cutoff2:
                    fr2 = self.sigma2 / r2
                    fr6 = fr2**3
                    U += 4.0*self.eps*fr6*(fr6 - 1.0) + self.eps
        return U

    def U_ext(self):
        U = 0.0
        for i in range(0,self.N):
            U += 0.5*self.omega*(self.x[i]**2 + self.y[i]**2 \
             + self.z[i]**2)
        return U

    def U_FENE(self):
        U = 0.0        
        for i in range(0,self.N-1):
            r2 = (self.x[i] - self.x[i+1])**2 +  \
                    (self.y[i] - self.y[i+1])**2 + \
                    (self.z[i] - self.z[i+1])**2
            U += 0.5*self.k_F*self.R0*np.log(1-r2/self.R0**2)
        return U

    def F_harm(self):
        F_x = np.zeros(self.N)
        F_y = np.zeros(self.N)
        F_z = np.zeros(self.N)

        if self.N == 1:
            return F_x, F_y, F_z
        
        F_x[0] = self.k_harm*(self.x[1] - self.x[0])
        F_y[0] = self.k_harm*(self.y[1] - self.y[0])
        F_z[0] = self.k_harm*(self.z[1] - self.z[0])

        F_x[self.N-1] = -self.k_harm*(self.x[self.N-1] - \
                self.x[self.N-2])
        F_y[self.N-1] = -self.k_harm*(self.y[self.N-1] - \
                self.y[self.N-2])
        F_z[self.N-1] = -self.k_harm*(self.z[self.N-1] - \
                self.z[self.N-2])

        for i in range(1,self.N-1):
            F_x[i] = self.k_harm*(self.x[i-1] - \
                     2.0*self.x[i] + self.x[i+1])
            F_y[i] = self.k_harm*(self.y[i-1] - \
                     2.0*self.y[i] + self.y[i+1])
            F_z[i] = self.k_harm*(self.z[i-1] - \
                     2.0*self.z[i] + self.z[i+1])
        
        return F_x, F_y, F_z

    def F_LJ(self):
        F_x = np.zeros(self.N)
        F_y = np.zeros(self.N)
        F_z = np.zeros(self.N)

        for i in range(0,self.N-1):
            for j in range(i+1,self.N):
                dx = self.x[i] - self.x[j]
                dy = self.y[i] - self.y[j]
                dz = self.z[i] - self.z[j]
                r2 = dx**2 + dy**2 + dz**2
                if r2 < self.cutoff2:
                    fr2 = self.sigma2 / r2
                    fr6 = fr2**3
                    fpr = 48.0*self.eps*(fr6*(fr6-0.5))/r2
                    F_x[i] += fpr*dx
                    F_y[i] += fpr*dy
                    F_z[i] += fpr*dz
                    F_x[j] -= fpr*dx
                    F_y[j] -= fpr*dy
                    F_z[i] -= fpr*dz
        return F_x, F_y, F_z
    
    def F_ext(self):
        F_x = np.zeros(self.N)
        F_y = np.zeros(self.N)
        F_z = np.zeros(self.N)

        for i in range(0,self.N):
            F_x[i] = -self.omega*self.x[i]
            F_y[i] = -self.omega*self.y[i]
            F_z[i] = -self.omega*self.z[i]

        return F_x, F_y, F_z

    def F_FENE(self):
        F_x = np.zeros(self.N)
        F_y = np.zeros(self.N)
        F_z = np.zeros(self.N)

        for i in range(0,self.N-1):
            dx = self.x[i+1] - self.x[i]
            dy = self.y[i+1] - self.y[i]
            dz = self.z[i+1] - self.z[i]
            r2 = dx**2 + dy**2 + dz**2
            fac = self.k_F/(1-r2/self.R0**2)
            F_x[i] += fac*dx
            F_y[i] += fac*dy
            F_z[i] += fac*dz
            F_x[i+1] -= fac*dx
            F_y[i+1] -= fac*dy
            F_z[i+1] -= fac*dz
        
        return F_x, F_y, F_z

    def F_tot(self):
        Fx_ext, Fy_ext, Fz_ext = self.F_ext()
        Fx_LJ, Fy_LJ, Fz_LJ = self.F_LJ()
        Fx_FENE, Fy_FENE, Fz_FENE = self.F_FENE()
        #Fx_harm, Fy_harm, Fz_harm = self.F_harm()
        return Fx_ext+Fx_LJ+Fx_FENE,Fy_ext+Fy_LJ+Fy_FENE, \
                Fz_ext+Fz_LJ+Fz_FENE
        #return Fx_harm+Fx_ext, Fy_harm+Fy_ext, Fz_harm+Fz_ext 

    def U_tot(self):
        U_kin = 0.0
        for i in range(0, self.N):
            U_kin += 0.5*(self.vx[i]**2 + self.vy[i]**2 + \
             self.vz[i]**2) 
        return self.U_LJ() +  self.U_FENE() + self.U_ext() + U_kin
        #return self.U_harm() + self.U_ext() + U_kin

    def update_cm(self):
        self.cmx = 0.0
        self.cmy = 0.0
        self.cmz = 0.0
        for i in range(0,self.N):
            self.cmx += self.x[i]
            self.cmy += self.y[i]
            self.cmz += self.z[i]
        self.cmx /= self.N
        self.cmy /= self.N
        self.cmz /= self.N

    def update_rg(self):
        self.update_cm()
        self.rgx = 0.0
        self.rgy = 0.0
        self.rgz = 0.0
        for i in range(0,self.N):
            self.rgx += (self.cmx - self.x[i])**2
            self.rgy += (self.cmy - self.y[i])**2
            self.rgz += (self.cmz - self.z[i])**2
        self.rgx /= self.N**2
        self.rgy /= self.N**2
        self.rgz /= self.N**2

    def leapfrog_update(self):
        for i in range(0,self.N):
            self.vhx[i] = self.vx[i] + self.fx[i]*self.dt/2.0
            self.vhy[i] = self.vy[i] + self.fy[i]*self.dt/2.0
            self.vhz[i] = self.vz[i] + self.fz[i]*self.dt/2.0
            self.x[i] += self.vhx[i] * self.dt
            self.y[i] += self.vhy[i] * self.dt
            self.z[i] += self.vhz[i] * self.dt
        
        self.fx, self.fy, self.fz = self.F_tot()
        for i in range(0,self.N):
            self.vx[i] = self.vhx[i] + self.fx[i]*self.dt/2.0
            self.vy[i] = self.vhy[i] + self.fy[i]*self.dt/2.0
            self.vz[i] = self.vhz[i] + self.fz[i]*self.dt/2.0

    def bbk_update(self):

        for i in range(0,self.N):
            self.vhx[i] = (1.0-0.5*self.gamma*self.dt)*self.vx[i] + \
                (self.fx[i] + self.rx[i]) * self.dt / 2.0
            self.vhy[i] = (1.0-0.5*self.gamma*self.dt)*self.vy[i] + \
                (self.fy[i] + self.ry[i]) * self.dt  / 2.0
            self.vhz[i] = (1.0-0.5*self.gamma*self.dt)*self.vz[i] + \
                (self.fz[i] + self.rz[i]) * self.dt  / 2.0

            self.x[i] += self.vhx[i] * self.dt
            self.y[i] += self.vhy[i] * self.dt
            self.z[i] += self.vhz[i] * self.dt

        self.fx, self.fy, self.fz = self.F_tot()        
        self.rx = np.random.normal(0,1,self.N)*self.mu
        self.ry = np.random.normal(0,1,self.N)*self.mu
        self.rz = np.random.normal(0,1,self.N)*self.mu

        for i in range(0,self.N):
            self.vx[i] = (self.vhx[i] + 0.5*(self.fx[i] + self.rx[i])\
                *self.dt)/(1.0+0.5*self.gamma*self.dt)
            self.vy[i] = (self.vhy[i] + 0.5*(self.fy[i] + self.ry[i])\
                *self.dt)/(1.0+0.5*self.gamma*self.dt)
            self.vz[i] = (self.vhz[i] + 0.5*(self.fz[i] + self.rz[i])\
                *self.dt)/(1.0+0.5*self.gamma*self.dt)


    def randomwalk_configuration(self, center):
        self.x[0] = self.y[0] = self.z[0] = 0.0
        for i in range(1,self.N):
            succeeded = False
            while not succeeded:
                theta = 1.0*np.pi*np.random.rand()
                phi = 2.0*np.pi*np.random.rand()
                self.x[i] = self.x[i-1] + np.sin(theta)*\
                                          np.cos(phi)
                self.y[i] = self.y[i-1] + np.sin(theta)*\
                                          np.sin(phi)
                self.z[i] = self.z[i-1] + np.cos(theta)

                succeeded = True 
                for j in range(0,i-1):
                    r2 = (self.x[i] - self.x[j])**2 +  \
                         (self.y[i] - self.y[j])**2 + \
                         (self.z[i] - self.z[j])**2
                    if r2 < self.cutoff2:
                        succeeded = False
        
        self.update_rg()
        self.update_cm()
        
        if center:
            for i in range(0,self.N):
                self.x[i] -= self.cmx
                self.y[i] -= self.cmy
                self.z[i] -= self.cmz
            

File: src/cpp/polymer.cpp
#include "polymer.hpp"
#include <iostream>
Polymer::Polymer(PolymerModel model, int N, float k_harm, float k_F, float R0,
                        float eps, float sigma, float omega,
                        float gamma, float T, float dt)
    : model(model), N(N), k_F(k_F), R0(R0), eps(eps),
      sigma(sigma), omega(omega), gamma(gamma),
      T(T), dt(dt),
      x(N, 0.0), y(N, 0.0), z(N, 0.0),
      vx(N, 0.0), vy(N, 0.0), vz(N, 0.0),
      vhx(N, 0.0), vhy(N, 0.0), vhz(N, 0.0),
      fx(N, 0.0), fy(N, 0.0), fz(N, 0.0),
      fx_harm(N, 0.0), fy_harm(N, 0.0), fz_harm(N, 0.0),
      fx_LJ(N, 0.0), fy_LJ(N, 0.0), fz_LJ(N, 0.0),
      fx_FENE(N, 0.0), fy_FENE(N, 0.0), fz_FENE(N, 0.0),
      fx_ext(N, 0.0), fy_ext(N, 0.0), fz_ext(N, 0.0),
      cmx(0.0), cmy(0.0), cmz(0.0),
      rgx(0.0), rgy(0.0), rgz(0.0),
      normal(0.0, 1.0), uniform(0.0, 1.0)
{
  mu = std::sqrt(2.0 * gamma * T / dt);
  PI = std::atan(2.0);
  sigma2 = sigma * sigma;
  cutoff2 = 1.25992104989 * sigma2;
  b1 = 1.0-0.5*gamma*dt;
  b2 = 1.0+0.5*gamma*dt;
  UpdateF();
};

void Polymer::ConfigureSAW(bool center)
{
  x[0] = 0.0;
  y[0] = 0.0;
  z[0] = 0.0;
  float theta, phi, dx, dy, dz, r2;

  for (int i = 1; i < N; ++i)
  {
    bool succeeded = false;
    while (!succeeded)
    {
      theta = PI * uniform(gen);
      phi = 2.0 * PI * uniform(gen);
      x[i] = x[i - 1] + std::sin(theta) * std::cos(phi);
      y[i] = y[i - 1] + std::sin(theta) * std::sin(phi);
      z[i] = z[i - 1] + std::cos(theta);

      succeeded = true;
      for (int j = 0; j < i - 1; ++j)
      {
        dx = x[i] - x[j];
        dy = y[i] - y[j];
        dz = z[i] - z[j];
        r2 = dx * dx + dy * dy + dz * dz;
        if (r2 < cutoff2)
          succeeded = false;
      }
    }
  }
  UpdateF();
  if (center)
  {
    Cm();
    for (int i = 0; i < N; ++i)
    {
      x[i] -= cmx;
      y[i] -= cmy;
      z[i] -= cmz;
    }
  }
};

void Polymer::Cm()
{
  cmx, cmy, cmz = 0.0;
  for (int i = 0; i < N; ++i)
  {
    cmx += x[i];
    cmy += y[i];
    cmz += z[i];
  }
  cmx /= N;
  cmy /= N;
  cmy /= N;
};

void Polymer::Rg()
{
  Cm();
  rgx, rgy, rgz = 0.0;
  for (int i = 0; i < N; ++i)
  {
    rgx += (cmx - x[i]) * (cmx - x[i]);
    rgy += (cmy - y[i]) * (cmy - y[i]);
    rgz += (cmz - z[i]) * (cmz - z[i]);
  }
  rgx /= N * N;
  rgy /= N * N;
  rgy /= N * N;
};

float Polymer::U_harm()
{
  float U = 0.0;
  float r2;
  for (int i = 0; i < N - 1; ++i)
  {
    r2 = (x[i] - x[i + 1]) * (x[i] - x[i + 1]) 
       + (y[i] - y[i + 1]) * (y[i] - y[i + 1]) 
       + (z[i] - z[i + 1]) * (z[i] - z[i + 1]);
    U += 0.5 * k_harm * r2;
  }
  return U;
};

float Polymer::U_LJ()
{
  float U = 0.0;
  for (int i = 0; i < N; ++i)
  {
    for (int j = i + 1; j < N; ++j)
    {
      r2 = (x[i] - x[i + 1]) * (x[i] - x[i + 1]) 
         + (y[i] - y[i + 1]) * (y[i] - y[i + 1]) 
         + (z[i] - z[i + 1]) * (z[i] - z[i + 1]);
      if (r2 < cutoff2)
      {
        fr6 = sigma2 * sigma2 * sigma2 / r2 / r2 / r2;
        U += 4.0 * eps * fr6 * (fr6 - 1.0) + eps;
      }
    }
  }
  return U;
};

float Polymer::U_FENE()
{
  float U = 0.0;
  for (int i = 0; i < N - 1; ++i)
  {
    r2 = (x[i] - x[i + 1]) * (x[i] - x[i + 1]) 
       + (y[i] - y[i + 1]) * (y[i] - y[i + 1]) 
       + (z[i] - z[i + 1]) * (z[i] - z[i + 1]);
    U += 0.5 * k_F * R0 * std::log(1 - r2 / R0 / R0);
  }
  return U;
};

float Polymer::U_ext()
{
  float U = 0.0;
  for (int i = 0; i < N; ++i)
  {
    U += 0.5 * omega * (x[i] * x[i] + y[i] * y[i] + z[i] * z[i]);
  }
  return U;
};

float Polymer::E_tot() {
  float E = 0.0;
  for (int i = 0; i < N; ++i) {
    E += 0.5*(vx[i]*vx[i] + vy[i]*vy[i] + vz[i]*vz[i]);
  }
  switch (model)
  {
    case Harmonic:
      E += U_harm() + U_ext();
      break;
    case SelfAvoiding:
      E += U_FENE() + U_LJ() + U_ext();
      break;
    default:
      break;
  }
  return E;
}

void Polymer::UpdateF_harm()
{
  fx_harm[0] = k_harm * (x[1] - x[0]);
  fy_harm[0] = k_harm * (y[1] - y[0]);
  fz_harm[0] = k_harm * (z[1] - z[0]);

  fx_harm[N - 1] = -k_harm * (x[N - 1] - x[N - 2]);
  fy_harm[N - 1] = -k_harm * (y[N - 1] - y[N - 2]);
  fz_harm[N - 1] = -k_harm * (z[N - 1] - z[N - 2]);

  for (int i = 1; i < N - 1; i++)
  {
    fx_harm[i] = k_harm * (x[i - 1] - 2.0 * x[i] + x[i + 1]);
    fy_harm[i] = k_harm * (y[i - 1] - 2.0 * y[i] + y[i + 1]);
    fz_harm[i] = k_harm * (z[i - 1] - 2.0 * z[i] + z[i + 1]);
  }
};

void Polymer::UpdateF_LJ()
{
  std::fill(fx_LJ.begin(), fx_LJ.end(), (float)0.0);
  std::fill(fy_LJ.begin(), fy_LJ.end(), (float)0.0);
  std::fill(fz_LJ.begin(), fz_LJ.end(), (float)0.0);

  for (int i = 0; i < N - 1; ++i)
  {
    for (int j = i + 1; j < N; ++j)
    {
      dx = x[i] - x[j];
      dy = y[i] - y[j];
      dz = z[i] - z[j];
      r2 = dx*dx + dy*dy + dz*dz;

      if(r2 < cutoff2) {
        fr6 = sigma2*sigma2*sigma2/r2/r2/r2;
        fpr = 48.0*eps*(fr6*(fr6-0.5))/r2;
        fx_LJ[i] += fpr*dx;
        fy_LJ[i] += fpr*dy;
        fz_LJ[i] += fpr*dz;        
        fx_LJ[j] -= fpr*dx;
        fy_LJ[j] -= fpr*dy;
        fz_LJ[j] -= fpr*dz;
      }
    }
  }
};

void Polymer::UpdateF_FENE() {
  std::fill(fx_FENE.begin(), fx_FENE.end(), (float)0.0);
  std::fill(fy_FENE.begin(), fy_FENE.end(), (float)0.0);
  std::fill(fz_FENE.begin(), fz_FENE.end(), (float)0.0);

  for (int i = 0; i < N - 1; ++i) {
      dx = x[i] - x[i+1];
      dy = y[i] - y[i+1];
      dz = z[i] - z[i+1];
      r2 = dx*dx + dy*dy + dz*dz;
      fac = k_F/(1-r2/R0/R0);
      fx_FENE[i] -= fac*dx;
      fy_FENE[i] -= fac*dy;
      fz_FENE[i] -= fac*dz;
      fx_FENE[i+1] += fac*dx;
      fy_FENE[i+1] += fac*dy;
      fz_FENE[i+1] += fac*dz;
  }
};

void Polymer::UpdateF_ext() {
    for (int i = 0; i < N; ++i) {
      fx_ext[i] = -omega*x[i];
      fy_ext[i] = -omega*y[i];
      fz_ext[i] = -omega*z[i];
    }
};

void Polymer::UpdateF() {
  switch (model)
  {
    case Harmonic:
      UpdateF_ext();
      UpdateF_harm();
      for(int i = 0; i < N; ++i) {
        fx[i] = fx_harm[i] + fx_ext[i];
        fy[i] = fy_harm[i] + fy_ext[i];
        fz[i] = fz_harm[i] + fz_ext[i];
      }
      break;
    case SelfAvoiding:
      UpdateF_ext();
      UpdateF_FENE();
      UpdateF_LJ();
      for(int i = 0; i < N; ++i) {
        fx[i] = fx_LJ[i] + fx_FENE[i] + fx_ext[i];
        fy[i] = fy_LJ[i] + fy_FENE[i] + fy_ext[i];
        fz[i] = fz_LJ[i] + fz_FENE[i] + fz_ext[i];
      }
      break;
    default:
      break;
  }

  if(T > 0.0) {
    for(int i = 0; i < N; ++i) {
      fx[i] += mu*normal(gen);
      fy[i] += mu*normal(gen);
      fz[i] += mu*normal(gen);
    }
  }
  
};

void Polymer::UpdateLeapFrog() {
  for(int i = 0; i < N; ++i) {
    vhx[i] = vx[i] + fx[i]*dt/2.0;
    vhy[i] = vy[i] + fy[i]*dt/2.0;
    vhz[i] = vz[i] + fz[i]*dt/2.0;
    x[i] += vhx[i]*dt;
    y[i] += vhy[i]*dt;
    z[i] += vhz[i]*dt;
  }
  UpdateF();
  for(int i = 0; i < N; ++i) {
    vx[i] = vhx[i] + fx[i]*dt/2.0;
    vy[i] = vhy[i] + fy[i]*dt/2.0;
    vz[i] = vhz[i] + fz[i]*dt/2.0;
  }
};

void Polymer::UpdateBBK() {

  for(int i = 0; i < N; ++i) {
    vhx[i] = b1*vx[i] + fx[i]*dt/2.0;
    vhy[i] = b1*vy[i] + fy[i]*dt/2.0;
    vhz[i] = b1*vz[i] + fz[i]*dt/2.0;
    
    x[i] += vhx[i]*dt;
    y[i] += vhy[i]*dt;
    z[i] += vhz[i]*dt;
  }
  UpdateF();
  for(int i = 0; i < N; ++i) {
    vx[i] = (vhx[i] + 0.5*fx[i]*dt)/b2;
    vy[i] = (vhy[i] + 0.5*fy[i]*dt)/b2;
    vz[i] = (vhz[i] + 0.5*fz[i]*dt)/b2;
  }
};  